offchain/authority-claimer/src/listener.rs

```diff
@@ -2,61 +2,152 @@
 // SPDX-License-Identifier: Apache-2.0 (see LICENSE)
 
 use async_trait::async_trait;
-use rollups_events::{BrokerConfig, DAppMetadata, RollupsClaim};
-use snafu::Snafu;
+use rollups_events::{
+    Broker, BrokerConfig, BrokerError, DAppMetadata, RollupsClaim,
+    RollupsClaimsStream, INITIAL_ID,
+};
+use snafu::ResultExt;
 use std::fmt::Debug;
 
-use crate::metrics::AuthorityClaimerMetrics;
-
-/// The `BrokerListener` listens for new claims from the broker.
-///
-/// The `listen` function should preferably yield to other processes while
-/// waiting for new messages (instead of busy-waiting).
+/// The `BrokerListener` listens for new claims from the broker
 #[async_trait]
 pub trait BrokerListener: Debug {
-    type Error: snafu::Error;
+    type Error: snafu::Error + 'static;
 
-    async fn listen(&self) -> Result<RollupsClaim, Self::Error>;
+    /// Listen to claims
+    async fn listen(&mut self) -> Result<RollupsClaim, Self::Error>;
 }
 
 // ------------------------------------------------------------------------------------------------
 // DefaultBrokerListener
 // ------------------------------------------------------------------------------------------------
 
 #[derive(Debug)]
-pub struct DefaultBrokerListener;
+pub struct DefaultBrokerListener {
+    broker: Broker,
+    stream: RollupsClaimsStream,
+    last_claim_id: String,
+}
 
-#[derive(Debug, Snafu)]
-pub enum DefaultBrokerListenerError {
-    Todo,
+#[derive(Debug, snafu::Snafu)]
+pub enum BrokerListenerError {
+    #[snafu(display("broker error"))]
+    BrokerError { source: BrokerError },
 }
 
 impl DefaultBrokerListener {
-    pub fn new(
-        _broker_config: BrokerConfig,
-        _dapp_metadata: DAppMetadata,
-        _metrics: AuthorityClaimerMetrics,
-    ) -> Result<Self, DefaultBrokerListenerError> {
-        todo!()
+    pub async fn new(
+        broker_config: BrokerConfig,
+        dapp_metadata: DAppMetadata,
+    ) -> Result<Self, BrokerError> {
+        tracing::trace!("Connecting to the broker ({:?})", broker_config);
+        let broker = Broker::new(broker_config).await?;
+        let stream = RollupsClaimsStream::new(&dapp_metadata);
+        let last_claim_id = INITIAL_ID.to_string();
+        Ok(Self {
+            broker,
+            stream,
+            last_claim_id,
+        })
     }
 }
 
 #[async_trait]
 impl BrokerListener for DefaultBrokerListener {
-    type Error = DefaultBrokerListenerError;
+    type Error = BrokerListenerError;
 
-    async fn listen(&self) -> Result<RollupsClaim, Self::Error> {
-        todo!()
+    async fn listen(&mut self) -> Result<RollupsClaim, Self::Error> {
+        tracing::trace!("Waiting for claim with id {}", self.last_claim_id);
+        let event = self
+            .broker
+            .consume_blocking(&self.stream, &self.last_claim_id)
+            .await
+            .context(BrokerSnafu)?;
+
+        Ok(event.payload.clone())
```

Comment:
```suggestion
        Ok(event.payload)
```

I believe this clone is unnecessary.